### PR TITLE
Issue# 114 - VesterTemplate.Tests.ps1 Performance Updates

### PR DESCRIPTION
**Pull request by Midacts**
I added jpsider's suggestion

- _Move the check for null items in the foreach ($scope) loop.
- #Only add valid ScopeObj's
- if (($ScopeObj.InventoryList -ne $NULL) -and ($ScopeObj.TestFiles -ne $NULL)){
- $Final += $ScopeObj
- }
- This will only add populated values to the new collection._

I did not add the other suggestions. I think the way where you get the object and loop through each test is how it is currently.

I'm not sure that we need these lines anymore:
```
    # The parent folder must be one of these names, to help with $Object scoping below
    # If adding here, also needs to be added to the switch below
    If ('vCenter|Datacenter|Cluster|DSCluster|Host|VM|Network' -notmatch $Scope) {
        Write-Warning "Skipping test $TestName. Use -Verbose for more details"
        Write-Verbose 'Test files should be in a folder with one of the following names:'
        Write-Verbose 'vCenter / Datacenter / Cluster / Host / VM / Network'
        Write-Verbose 'This helps Vester determine which inventory object(s) to use during the test.'
        # Use continue to skip this test and go to the next loop iteration
        continue
    }
```

and

```
    If (($Final | Where-Object { $_.Scope -eq $Scope }).InventoryList -eq $null) {
        Write-Verbose "No objects found in scope $Scope, skipping tests"
        # Use continue to skip this test and go to the next loop iteration
        continue
    }    
```

mainly because of lines 51-53

### CODE REVIEW

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 1]
```diff
@@ -53,63 +85,57 @@ ForEach ($Test in $TestFiles) {
...
-                'Datacenter' {$Datacenter}
-                'Cluster'    {$Datacenter | Get-Cluster -Name $cfg.scope.cluster}
-                'DSCluster'  {$Datacenter | Get-DatastoreCluster -Name $cfg.scope.dscluster}
-                'Host'       {$Datacenter | Get-Cluster -Name $cfg.scope.cluster | Get-VMHost -Name $cfg.scope.host}
-                'VM'         {$Datacenter | Get-Cluster -Name $cfg.scope.cluster | Get-VM -Name $cfg.scope.vm}
-                'Network'    {$Datacenter | Get-VDSwitch -Name $cfg.scope.vds}
-            }
-        } #If Desired
+    # Loops through each object in the inventory list for the specific scope
+    foreach($Object in $Inventory)
```
Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: I believe these two loops need to be switched in order to maintain the current behavior. Check out these two images that show the difference in the console output.
current
proposed

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 2]
```diff
@@ -53,63 +85,57 @@ ForEach ($Test in $TestFiles) {
         }
     }
 
-    Describe -Name "$Scope Configuration: $TestName" -Fixture {
-        # Pull in $Title/$Description/$Desired/$Type/$Actual/$Fix from the test file
-        . $Test
+    If (($Final | Where-Object { $_.Scope -eq $Scope }).InventoryList -eq $null) {
```
Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: I agree that this code is not needed, since you are checking for null values before you add the items to the $final collection.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 3]
```diff
@@ -22,18 +22,50 @@ Param(
...
+foreach($Scope in $Final.Scope)
+{
+    # Pulling the inventory and test files for this scope
+    $Inventory = ($Final | Where-Object { $_.Scope -eq $Scope }).InventoryList
+    $Tests = ($Final | Where-Object { $_.Scope -eq $Scope }).TestFiles
 
     # The parent folder must be one of these names, to help with $Object scoping below
     # If adding here, also needs to be added to the switch below
-    If ($Scope -notmatch 'vCenter|Datacenter|Cluster|DSCluster|Host|VM|Network') {
+    If ('vCenter|Datacenter|Cluster|DSCluster|Host|VM|Network' -notmatch $Scope) {
```
Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: I think you still need this, it is the only verification that the items that are found are supported by vester.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 4]
```diff
@@ -22,18 +22,50 @@ Param(
...
         Write-Warning "Skipping test $TestName. Use -Verbose for more details"
         Write-Verbose 'Test files should be in a folder with one of the following names:'
-        Write-Verbose 'vCenter / Datacenter / Cluster / DSCluster / Host / VM / Network'
+        Write-Verbose 'vCenter / Datacenter / Cluster / Host / VM / Network'
```
Comment (jonneedham) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Any reason we got rid of the DSCluster here in Write-Verbose?

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 1 above)
Comment (jonneedham) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: If we start the master loop by each $Object in $Inventory, the Pester output will be sorted by $Object (e.g. VM), and then we'll see output for each test run against that $Object.  Ideally, Pester output will be sorted by test, and we'll loop through each object that applies to the test.  This makes going back and looking through the output more easy to understand / logical for the end user.  If grouped by test, it's easier to see "Oh, shoot...that one VM advanced setting isn't applied to half of my VMs, wtf?" than it is to see when grouped by $Object (in my opinion...).

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 5]
```diff
@@ -22,18 +22,50 @@ Param(
...
+    Write-Verbose "Processing $Scope"
+    Remove-Variable InventoryList -ErrorAction SilentlyContinue # Makes sure the variable is always fresh
+    # Use $Scope (parent folder) to get the correct objects to test against
+    # If changing values here, update the "$Scope -notmatch" test below as well
+    $InventoryList = switch ($Scope) {
+        'vCenter'    {$cfg.vcenter.vc}
+        'Datacenter' {$Datacenter}
+        'Cluster'    {$Datacenter | Get-Cluster -Name $cfg.scope.cluster}
+        'DSCluster'  {$Datacenter | Get-DatastoreCluster -Name $cfg.scope.dscluster}
+        'Host'       {$Datacenter | Get-Cluster -Name $cfg.scope.cluster | Get-VMHost -Name $cfg.scope.host}
```
Comment (jonneedham) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Is it possible to use $Cluster | Get-VMHost instead of getting the cluster a second time?  Not a huge deal, and we can leave this code alone for this PR if necessary.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 6]
```diff
@@ -22,18 +22,50 @@ Param(
...
+        'VM'         {$Datacenter | Get-Cluster -Name $cfg.scope.cluster | Get-VM -Name $cfg.scope.vm}
```
Comment (jonneedham) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Same as line 41.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 5 above)
Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: technically you are not getting it a second time. I do think that a new issue needs to be submitted to exclude dependencies.

Comment (jonneedham) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Not sure I follow...you're running $Datacenter | Get-Cluster multiple times.  How is that not getting the cluster multiple times?  We aren't referencing an object contained within $Datacenter...we're doing another Get call to the vCenter right?  I may be missing something.

Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: the problem is that it's a switch, and we are not storing a $cluster anywhere. so in order to remove the multiple 'gets' we would need to store the common items. So I think that would improve performance, but I think that it should probably be a new issue, not 114.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 4 above)
Comment (Midacts) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Thanks for catching that. DSCluster was missing by mistake.
I have added it back in.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 1 above)
Comment (Midacts) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: I think that is the same point that jpsider points out below with his pictures. We should just need to swap the foreach loops. ill test it out and make a need commit.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 5 above)
Comment (Midacts) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: It seems to me that we could just move $Cluster to the top like we did for $Datacenter.
Since $Datacenter and $Cluster are used by alot of other scopes. (VM, host, etc).
From what I remember, it should just be this file and New-VesterConfig.ps1 that are running these kinds of 'Gets'

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 3 above)
Comment (Midacts) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: I agree. i just wanted the teams opinion before i just axed anything.
I'll keep this one in there.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 2 above)
Comment (Midacts) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: I think so too. Thanks for verifying.
I removed this code in it's own commit.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 1 above)
Comment (Midacts) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: That makes sense. I was under the impression that the way i had it was the way it was currently?
host1 -> loops through all the tests
host2 -> loops through all the tests, etc

with this change:
each test -> each host gets the test run against it -> then it moves on to the next test.

I made the switch, but either way it is a super easy fix.

**Comment by Midacts** [comment]
All the suggested changes have went through and I have tested them.
Let me know if you guys see anything else.

Thanks again for your review and suggestions.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 7]
```diff
@@ -51,65 +83,55 @@ ForEach ($Test in $TestFiles) {
...
-                    }
-                } #Try/Catch
-            } #It
-        } #ForEach Object
-    } #Describe
-} #ForEach Test
+                    } #Try/Catch
+                } #It
+            } #Describe                    
+        }#Foreach Testfiles
```
Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: super minor, but these could be reversed and cleaned up, or removed

**Comment by Midacts** [comment]
That was a good find (I always miss stuff like that).

**Comment by brianbunke** [comment]
First, thanks to everyone for this effort! I definitely apologize for this sitting so long. It's a very visible change, so I wanted to give it a test drive myself, and I'm just getting to that now.

I haven't reviewed every change to the code yet, but I have one suggestion that I'm adding to my new review.

In the PR as is, the many Get- calls are resolved (yay!), but it suffers from a different speed problem. Within each test, a new Describe block is created for every single object, and _then_ the "is the config value null?" check is performed. This makes null value tests take far longer, because if a single VM test's config value is null, it still has to create a new Describe block and then skip for every single VM you have.

File: Vester/Private/Template/VesterTemplate.Tests.ps1 [hunk 8]
```diff
@@ -51,65 +83,55 @@ ForEach ($Test in $TestFiles) {
...
-                'DSCluster'  {$Datacenter | Get-DatastoreCluster -Name $cfg.scope.dscluster}
-                'Host'       {$Datacenter | Get-Cluster -Name $cfg.scope.cluster | Get-VMHost -Name $cfg.scope.host}
-                'VM'         {$Datacenter | Get-Cluster -Name $cfg.scope.cluster | Get-VM -Name $cfg.scope.vm}
-                'Network'    {$Datacenter | Get-VDSwitch -Name $cfg.scope.vds}
-            }
-        } #If Desired
+        # Loops through each object in the inventory list for the specific scope.
+        # It runs one test at a time against each $Object and moves onto the next test.
+        foreach($Object in $Inventory)
+        {
```
Comment (brianbunke) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: 92-95 should move below 96-108. This ensures that for each test, only one Describe block is created for all objects, and the `$Desired -eq $null` check happens once for each test (instead of once for each object inside each test).

**Comment by brianbunke** [comment]
I should note, the combination of changes in this PR (including the suggestion in my review) resulted in my benchmark test just now going from 150 seconds down to 65.

Seems pretty good 👍

File: Vester/Private/Template/VesterTemplate.Tests.ps1 (hunk 8 above)
Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Do you believe this is outside the scope of the ticket? Although the title is fairly wide open. I believe that @Midacts  wanted to make changes and not even include the test if the config value was null. But that would happen before this file was executed.

Comment (brianbunke) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: My suggestion is not outside scope, as it is how Vester currently behaves. It helps ensure that this PR adheres to current standards.

I agree that ignoring the test altogether if the scope is null is outside the scope of this PR, but we're talking about two slightly different things.

Comment (jpsider) [inline Vester/Private/Template/VesterTemplate.Tests.ps1]: Sounds fair!

**Comment by brianbunke** [comment]
I believe the `ForEach ($Object in $Inventory) {` should move below the `$null` check at line 108. This PR right now, lines 92-108:

- for each Vester test,
- for each VMware inventory object,
- create a unique Describe block,
- and check if the $cfg value for this test is `$null`

What matches current behavior, and what I believe to be correct:

- for each Vester test,
- create one Describe block,
- check if the $cfg value for this test is `$null`,
- and then loop through each VMware inventory object

Moving the Describe block & `$cfg` value check up really speeds up tests where there is a `$null` value. I see this in action when many VM hardening tests run against >100 VMs.

**Comment by Midacts** [comment]
I added the requested changes.
Let me know if you see anything else that needs to change

**Comment by brianbunke** [comment]
Kudos to @Midacts for sticking with this, and much love to @jpsider and @jonneedham for the reviews. As far as I'm concerned, Vester just doubled in speed. 🍾

:shipit:

**Comment by Midacts** [comment]
No problem. I am glad we were all able to work together to find the best solution.
Glad to be a part of the team.
